rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,417 @@
+
+The backend will also create an entry in the upload table using the dataset_id as the key, and set the status to waiting.
+
+#### 4. Queue
+
+The queue stores the links that need to be processed. It buffers the downstream servers from download requests. This
+allows us to spin up servers as needed. It also gives us the ability to retry a download if one of the servers to fails
+upload the file.
+
+##### Why Might the Upload Fail
+
+- The downloading server ran out of storage.
+- The downloading server crashed.
+- DropBox experience an outage
+- The user has reached their [daily bandwidth allowance](https://www.dropbox.com/plans?trigger=nr) from their DropBox. [more info](https://help.dropbox.com/files-folders/share/banned-links).
+
+#### 5. Start an Instance to Download File
+
+An EC2 instance, docker container, or AWS lambda can be used to complete this task. Once a computing resource is spawned,
+it will change the upload status to uploading.
+
+#### 6. Stream File
+
+The shared link is used to stream the file from one cloud to another. The most time will be spent streaming the file
+from the DropBox to S3. DropBox offers a way to validate the integrity of a download
+by providing a hash of the data, for example [Dropbox Content Hash](https://www.dropbox.com/developers/reference/content-hash).
+As the data is streaming from the CSP, the hash of the file will be calculated as it arrives, and again as it's
+uploaded to S3. The [Content-MD5 header](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax)
+will be used when uploading to S3 to verify the integrity of each chunk. The AWS SDK will be useful in simplifying the
+upload to S3. For calculating the checksum while streaming, [checksumming_io](https://github.com/HumanCellAtlas/checksumming_io)
+can be used.
+
+Once the upload is complete, the download job will be removed from the queue. If the final hash does not match, then
+the upload process will be restarted. The retry logic will be an exponential backoff, with a max retry of 5 times.
+
+While streaming the file, the computing resource will periodically check if the upload has been canceled using the upload
+table. If the upload status is cancel pending, the computing resource will stop the stream, delete any data that has been uploaded,
+mark the job as complete in the queue, and set the status of the upload to cancelled.
+
+Where possible, the progress of the upload will be calculated and updated in the upload table.
+
+#### 7. Retry
+
+If the upload fails for any reason, the computing resource will mark the status of the upload as waiting, and return the upload
+job to the queue.
+
+#### 8. Failed Uploads
+
+If an upload fails enough times, it will be removed from the download queue. The upload status will change to failed.
```

Comment:
RE _enough times_ - I would replace with the precise number or specify something like _maximum number of retries allowed by the service_

Reply:
To avoid repeating myself I will add a link back to the retry section.